byterun/fail.c

```diff
@@ -81,68 +81,97 @@ CAMLexport void caml_raise_with_string(value tag, char const *msg)
+   while reading the initial value of [caml_global_data].
+
+   We check against this issue here in byterun/fail.c instead of
+   byterun/intern.c. Having the check here means that these calls will
+   be slightly slower for all bytecode programs (not just the calls
+   coming from intern). Because intern.c is shared between byterun/
+   and asmrun/, putting checks there would slow do input_value for
+   natively-compiled programs that do not need these checks.
+*/
+static void check_global_data(char const *exception_name)
 {
   if (caml_global_data == 0) {
-    fprintf(stderr, "Fatal error: exception Failure(\"%s\")\n", msg);
+    fprintf(stderr, "Fatal error: exception %s\n", exception_name);
     exit(2);
   }
-  caml_raise_with_string(Field(caml_global_data, FAILURE_EXN), msg);
 }
 
-CAMLexport void caml_invalid_argument (char const *msg)
+static void check_global_data_param(char const *exception_name, char const *msg)
 {
   if (caml_global_data == 0) {
-    fprintf(stderr, "Fatal error: exception Invalid_argument(\"%s\")\n", msg);
+    fprintf(stderr, "Fatal error: exception %s(\"%s\")\n", exception_name, msg);
     exit(2);
   }
+}
+
+CAMLexport void caml_failwith (char const *msg)
+{
+  check_global_data_param("Failure", msg);
+  caml_raise_with_string(Field(caml_global_data, FAILURE_EXN), msg);
+}
+
+CAMLexport void caml_invalid_argument (char const *msg)
+{
+  check_global_data_param("Invalid_argument", msg);
   caml_raise_with_string(Field(caml_global_data, INVALID_EXN), msg);
 }
 
 CAMLexport void caml_array_bound_error(void)
 {
+  check_global_data_param("Invalid_argument", "index out of bounds");
   caml_invalid_argument("index out of bounds");
 }
 
 CAMLexport void caml_raise_out_of_memory(void)
 {
+  check_global_data("Out_of_memory");
   caml_raise_constant(Field(caml_global_data, OUT_OF_MEMORY_EXN));
 }
 
 CAMLexport void caml_raise_stack_overflow(void)
 {
+  check_global_data("Stack_overflow");
   caml_raise_constant(Field(caml_global_data, STACK_OVERFLOW_EXN));
 }
 
 CAMLexport void caml_raise_sys_error(value msg)
 {
+  check_global_data_param("Sys_error", String_val(msg));
   caml_raise_with_arg(Field(caml_global_data, SYS_ERROR_EXN), msg);
 }
 
 CAMLexport void caml_raise_end_of_file(void)
 {
+  check_global_data("End_of_file");
   caml_raise_constant(Field(caml_global_data, END_OF_FILE_EXN));
 }
 
 CAMLexport void caml_raise_zero_divide(void)
 {
+  check_global_data("Division_by_zero");
   caml_raise_constant(Field(caml_global_data, ZERO_DIVIDE_EXN));
 }
 
 CAMLexport void caml_raise_not_found(void)
 {
+  check_global_data("Not_found");
   caml_raise_constant(Field(caml_global_data, NOT_FOUND_EXN));
 }
 
 CAMLexport void caml_raise_sys_blocked_io(void)
 {
+  check_global_data("Sys_blocked_io");
   caml_raise_constant(Field(caml_global_data, SYS_BLOCKED_IO));
 }
 
 int caml_is_special_exception(value exn) {
+  if (caml_global_data == 0) {
+    fprintf(stderr, "Fatal error: caml_is_special_exception\n");
+    exit(2);
+  }
```

Comment:
It would be better to return 0 in case `caml_global_data == 0`.  Reason: `caml_is_special_exception` is used only in `caml_format_exception` to produce a more readable textual representation of some exceptions.  It is better to fall back to the general, less readable representation than to abort with a fatal error.